src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5042,6 +5042,288 @@ class StubGenerator: public StubCodeGenerator {
     return (address) start;
   }
 
+  void adler32_process_bytes(Register buff, Register s1, Register s2, Register count,
+    VectorRegister vtable, VectorRegister vzero, VectorRegister vbytes, VectorRegister vs1acc, VectorRegister *vs2acc,
+    Register temp0, Register temp1, Register temp2, VectorRegister vtemp1, VectorRegister vtemp2,
+    int step, Assembler::LMUL LMUL) {
+
+    assert((LMUL == Assembler::m4 && step == 64) || (LMUL == Assembler::m2 && step == 32) ||
+      (LMUL == Assembler::m1 && step == 16), "LMUL should be aligned with step: m4 and 64, m2 and 32 or m1 and 16");
+    // Below is function for calculating Adler32 checksum with 64-, 32- or 16-byte step. LMUL=m4, m2 or m1 is used.
+    // The results are in v12, v13, ..., v22, v23. Example below is for 64-byte step case.
+    // We use b1, b2, ..., b64 to denote the 64 bytes loaded in each iteration.
+    // In non-vectorized code, we update s1 and s2 as:
+    //   s1 <- s1 + b1
+    //   s2 <- s2 + s1
+    //   s1 <- s1 + b2
+    //   s2 <- s2 + b1
+    //   ...
+    //   s1 <- s1 + b64
+    //   s2 <- s2 + s1
+    // Putting above assignments together, we have:
+    //   s1_new = s1 + b1 + b2 + ... + b64
+    //   s2_new = s2 + (s1 + b1) + (s1 + b1 + b2) + ... + (s1 + b1 + b2 + ... + b64) =
+    //          = s2 + s1 * 64 + (b1 * 64 + b2 * 63 + ... + b64 * 1) =
+    //          = s2 + s1 * 64 + (b1, b2, ... b64) dot (64, 63, ... 1)
+
+    __ mv(count, step);
+    // Load data
+    __ vsetvli(temp0, count, Assembler::e8, LMUL);
+    __ vle8_v(vbytes, buff);
+    __ addi(buff, buff, step);
+
+    // Upper bound reduction sum for s1_new:
+    // 0xFF * 64 = 0x3FC0, so:
+    // 1. Need to do vector-widening reduction sum
+    // 2. It is safe to perform sign-extension during vmv.x.s with 16-bits elements
+    __ vwredsumu_vs(vs1acc, vbytes, vzero);
+    // Multiplication for s2_new
+    __ vwmulu_vv(vs2acc[0], vtable, vbytes);
+
+    // s2 = s2 + s1 * log2(step)
+    __ slli(temp1, s1, exact_log2(step));
+    __ add(s2, s2, temp1);
+
+    // MaxVectorSize == 16: Remainder for vwmulu.vv is in successor of vs2acc[0] group
+    VectorRegister vs2acc_successor = vs2acc[2];
+    if (step == 32)
+      vs2acc_successor = vs2acc[1];
+    else if (step == 16)
+      vs2acc_successor = vs2acc[0]->successor();
+
+    // Summing up calculated results for s2_new
+    __ vsetvli(temp0, count, Assembler::e16, LMUL);
+    // Upper bound for reduction sum:
+    // 0xFF * (64 + 63 + ... + 2 + 1) = 0x817E0 max for whole register group, so:
+    // 1. Need to do vector-widening reduction sum
+    // 2. It is safe to perform sign-extension during vmv.x.s with 32-bits elements
+    __ vwredsumu_vs(vtemp1, vs2acc[0], vzero);
```

Comment:
Sounds good, thanks! As far as I can see it is safe to do it for any `MaxVectorSize`, so that the code will be even simplier. Please correct me if I'm wrong

Reply:
Yeah, it will be more simpler if you use it for any MaxVectorSize. But to my surprise, it will bring negative performane impact when I test on BananaPi-F3. I am not sure whether it's issue across the boards. Maybe you can give it a try on your hardware to see?

Reply:
Hi! Sorry for such a late reply. Here are the results on Kendryte K230 for both options.
For stable `MaxVectorSize = 32`:
```
Benchmark                              (count)   Mode  Cnt     Score   Error   Units
Adler32.TestAdler32.testAdler32Update       64  thrpt   25  7253.108 ? 7.362  ops/ms
Adler32.TestAdler32.testAdler32Update      128  thrpt   25  5849.102 ? 9.263  ops/ms
Adler32.TestAdler32.testAdler32Update      256  thrpt   25  4220.237 ? 4.695  ops/ms
Adler32.TestAdler32.testAdler32Update      512  thrpt   25  2717.498 ? 3.531  ops/ms
Adler32.TestAdler32.testAdler32Update     1024  thrpt   25  1585.664 ? 2.455  ops/ms
Adler32.TestAdler32.testAdler32Update     2048  thrpt   25   865.849 ? 0.717  ops/ms
Adler32.TestAdler32.testAdler32Update     5012  thrpt   25   372.534 ? 0.293  ops/ms
Adler32.TestAdler32.testAdler32Update     8192  thrpt   25   231.532 ? 0.306  ops/ms
Adler32.TestAdler32.testAdler32Update    16384  thrpt   25   117.180 ? 0.157  ops/ms
Adler32.TestAdler32.testAdler32Update    32768  thrpt   25    55.013 ? 0.152  ops/ms
Adler32.TestAdler32.testAdler32Update    65536  thrpt   25    25.604 ? 0.126  ops/ms
```

and for different `MaxVectorSize`:
```
Benchmark                              (count)   Mode  Cnt     Score   Error   Units
Adler32.TestAdler32.testAdler32Update       64  thrpt   25  7239.485 ? 8.705  ops/ms
Adler32.TestAdler32.testAdler32Update      128  thrpt   25  5836.018 ? 9.489  ops/ms
Adler32.TestAdler32.testAdler32Update      256  thrpt   25  4212.986 ? 4.596  ops/ms
Adler32.TestAdler32.testAdler32Update      512  thrpt   25  2712.742 ? 3.114  ops/ms
Adler32.TestAdler32.testAdler32Update     1024  thrpt   25  1583.161 ? 2.374  ops/ms
Adler32.TestAdler32.testAdler32Update     2048  thrpt   25   864.321 ? 0.870  ops/ms
Adler32.TestAdler32.testAdler32Update     5012  thrpt   25   371.964 ? 0.463  ops/ms
Adler32.TestAdler32.testAdler32Update     8192  thrpt   25   231.092 ? 0.328  ops/ms
Adler32.TestAdler32.testAdler32Update    16384  thrpt   25   116.995 ? 0.189  ops/ms
Adler32.TestAdler32.testAdler32Update    32768  thrpt   25    54.923 ? 0.075  ops/ms
Adler32.TestAdler32.testAdler32Update    65536  thrpt   25    24.864 ? 0.618  ops/ms
```

Reply:
Hi! Sorry for such a late reply. Here are the results on Banana-Pi. 
Numbers for different `vsetvli` depending on the `MaxVectorSize` (= 32 by default on Banana-Pi):
```
Benchmark                              (count)   Mode  Cnt     Score     Error   Units
Adler32.TestAdler32.testAdler32Update       64  thrpt   25  7564.243 ± 103.398  ops/ms
Adler32.TestAdler32.testAdler32Update      128  thrpt   25  5691.398 ±  72.201  ops/ms
Adler32.TestAdler32.testAdler32Update      256  thrpt   25  3831.476 ±  32.985  ops/ms
Adler32.TestAdler32.testAdler32Update      512  thrpt   25  2325.933 ±  10.056  ops/ms
Adler32.TestAdler32.testAdler32Update     1024  thrpt   25  1311.345 ±   3.337  ops/ms
Adler32.TestAdler32.testAdler32Update     2048  thrpt   25   697.412 ±   0.906  ops/ms
Adler32.TestAdler32.testAdler32Update     5012  thrpt   25   294.467 ±   0.131  ops/ms
Adler32.TestAdler32.testAdler32Update     8192  thrpt   25   182.197 ±   0.075  ops/ms
Adler32.TestAdler32.testAdler32Update    16384  thrpt   25    92.100 ±   0.075  ops/ms
Adler32.TestAdler32.testAdler32Update    32768  thrpt   25    45.242 ±   0.006  ops/ms
Adler32.TestAdler32.testAdler32Update    65536  thrpt   25    22.878 ±   0.008  ops/ms
```

and for stable `vsetvli` with `LMULx2` for any `MaxVectorSize`:
```
Benchmark                              (count)   Mode  Cnt     Score     Error   Units
Adler32.TestAdler32.testAdler32Update       64  thrpt   25  7439.400 ± 128.059  ops/ms
Adler32.TestAdler32.testAdler32Update      128  thrpt   25  5662.910 ±  73.459  ops/ms
Adler32.TestAdler32.testAdler32Update      256  thrpt   25  3831.516 ±  33.050  ops/ms
Adler32.TestAdler32.testAdler32Update      512  thrpt   25  2345.353 ±   9.946  ops/ms
Adler32.TestAdler32.testAdler32Update     1024  thrpt   25  1305.337 ±   3.294  ops/ms
Adler32.TestAdler32.testAdler32Update     2048  thrpt   25   695.615 ±   0.775  ops/ms
Adler32.TestAdler32.testAdler32Update     5012  thrpt   25   294.332 ±   0.132  ops/ms
Adler32.TestAdler32.testAdler32Update     8192  thrpt   25   182.178 ±   0.086  ops/ms
Adler32.TestAdler32.testAdler32Update    16384  thrpt   25    92.125 ±   0.032  ops/ms
Adler32.TestAdler32.testAdler32Update    32768  thrpt   25    45.238 ±   0.030  ops/ms
Adler32.TestAdler32.testAdler32Update    65536  thrpt   25    22.880 ±   0.008  ops/ms
```

Reply:
So I suppose it is safe to stay on conditionless `vsetvli`?

Reply:
Hmm ... JMH data on my banana-pi (running a 6.1.15 kernel from the vendor) is kind of different from yours for these two approaches.

1. __ vsetvli(temp0, count, Assembler::e16, LMUL) for (MaxVectorSize > 16)
```
Benchmark                      (count)   Mode  Cnt     Score     Error   Units
TestAdler32.testAdler32Update       64  thrpt   25  7364.310 ± 103.256  ops/ms
TestAdler32.testAdler32Update      128  thrpt   25  5651.856 ±  71.376  ops/ms
TestAdler32.testAdler32Update      256  thrpt   25  3803.744 ±  18.320  ops/ms
TestAdler32.testAdler32Update      512  thrpt   25  2324.802 ±   8.553  ops/ms
TestAdler32.testAdler32Update     1024  thrpt   25  1306.936 ±   4.027  ops/ms
TestAdler32.testAdler32Update     2048  thrpt   25   696.408 ±   1.925  ops/ms
TestAdler32.testAdler32Update     5012  thrpt   25   294.126 ±   0.644  ops/ms
TestAdler32.testAdler32Update     8192  thrpt   25   182.142 ±   0.048  ops/ms
TestAdler32.testAdler32Update    16384  thrpt   25    92.007 ±   0.253  ops/ms
TestAdler32.testAdler32Update    32768  thrpt   25    45.190 ±   0.158  ops/ms
TestAdler32.testAdler32Update    65536  thrpt   25    22.873 ±   0.014  ops/ms
```

2. __ vsetvli(temp0, count, Assembler::e16, LMULx2) for (MaxVectorSize == 16)
```
Benchmark                      (count)   Mode  Cnt     Score    Error   Units
TestAdler32.testAdler32Update       64  thrpt   25  7683.759 ± 92.761  ops/ms
TestAdler32.testAdler32Update      128  thrpt   25  6226.934 ± 71.597  ops/ms
TestAdler32.testAdler32Update      256  thrpt   25  4409.333 ± 27.677  ops/ms
TestAdler32.testAdler32Update      512  thrpt   25  2813.737 ±  5.570  ops/ms
TestAdler32.testAdler32Update     1024  thrpt   25  1635.601 ±  1.207  ops/ms
TestAdler32.testAdler32Update     2048  thrpt   25   891.615 ±  0.999  ops/ms
TestAdler32.testAdler32Update     5012  thrpt   25   382.035 ±  0.255  ops/ms
TestAdler32.testAdler32Update     8192  thrpt   25   237.338 ±  0.282  ops/ms
TestAdler32.testAdler32Update    16384  thrpt   25   120.517 ±  0.044  ops/ms
TestAdler32.testAdler32Update    32768  thrpt   25    58.957 ±  0.059  ops/ms
TestAdler32.testAdler32Update    65536  thrpt   25    29.881 ±  0.009  ops/ms
```

Reply:
This is weird. Ok, let's be more conservative with two-`LMUL` approach then

Reply:
I also optimized generation of `vtable_32`/`vtable_16` depending on the `MaxVectorSize`.